### PR TITLE
Fix cassandra tests intermittent failures

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSession.java [hunk 1]
```diff
@@ -94,35 +94,35 @@ public Session load(String key)
...
-        return executeWithSession(schemaName, session -> session.getCluster().getMetadata().getReplicas(schemaName, partitionKey));
+        return executeWithSession(session -> session.getCluster().getMetadata().getReplicas(schemaName, partitionKey));
     }
 
-    private Session getSession(String schema)
+    private Session getSession()
     {
         try {
-            return sessionBySchema.get(schema);
+            return sessionBySchema.get("");
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSession.java]: Are you positive we use the fully qualified tablename in all queries? (see https://docs.datastax.com/en/drivers/java/2.0/com/datastax/driver/core/Session.html)

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/MockCassandraSession.java [hunk 2]
```diff
@@ -124,35 +116,18 @@ public CassandraTable getTable(SchemaTableName tableName)
...
-        if (throwException) {
-            throw new IllegalStateException();
-        }
-
-        return super.getPartitions(table, filterPrefix);
+        throw new UnsupportedOperationException();
     }
 
     @Override
-    protected List<Row> queryPartitionKeys(CassandraTable table, List<Object> filterPrefix)
```
Comment (maciejgrzybek) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/MockCassandraSession.java]: This probably should be a separate commit, to ensure that this commit is purely an automatic extraction of interface.
I won't insist if it's too much hassle to split it.

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/CassandraQueryRunner.java [hunk 3]
```diff
@@ -47,13 +46,12 @@ public static synchronized DistributedQueryRunner createCassandraQueryRunner()
 
         queryRunner.installPlugin(new CassandraPlugin());
         queryRunner.createCatalog("cassandra", "cassandra", ImmutableMap.of(
-                "cassandra.contact-points", "localhost",
-                "cassandra.native-protocol-port", "9142",
+                "cassandra.contact-points", EmbeddedCassandra.getHost(),
```
Comment (maciejgrzybek) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/CassandraQueryRunner.java]: Shouldn't it be a separate commit?

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java [hunk 4]
```diff
@@ -13,20 +13,90 @@
...
+                .addContactPointsWithPorts(ImmutableList.of(
+                        new InetSocketAddress(
+                                EmbeddedCassandraServerHelper.getHost(),
+                                EmbeddedCassandraServerHelper.getNativeTransportPort())))
+                .build();
+        try {
+            checkConnectivity(cluster);
+        }
+        catch (RuntimeException e) {
+            cluster.close();
```
Comment (maciejgrzybek) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java]: Put this in `finally` instead.

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/CassandraTestingUtils.java [hunk 5]
```diff
@@ -44,7 +44,7 @@ private CassandraTestingUtils() {}
     public static void createTestTables(Cluster cluster, String keyspace, Date date)
     {
         try (Session session = cluster.connect()) {
-            createOrReplaceKeyspace(session, keyspace);
+            createKeyspace(session, keyspace);
         }
         try (Session session = cluster.connect(keyspace)) {
```
Comment (maciejgrzybek) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/CassandraTestingUtils.java]: And catch with `propagate` in case there's a problem with creating a keyspace (due to e.g. name clash). It should never happen but given it's our assertion here, we should not hide this.

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java [hunk 6]
```diff
@@ -85,13 +125,13 @@ public static synchronized CassandraSession getSession()
     public static synchronized String getHost()
     {
         checkIsInitialized();
-        return "127.0.0.1";
+        return HOST;
```
Comment (maciejgrzybek) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java]: It should be in the commit changing `EmbeddedCassandra`, not here.

File: presto-cassandra/src/test/resources/cu-cassandra.yaml [hunk 7]
```diff
@@ -94,10 +94,10 @@ partitioner: org.apache.cassandra.dht.Murmur3Partitioner
 
 # directories where Cassandra should store data on disk.
 data_file_directories:
-    - target/embeddedCassandra/data
+    - ${data_directory}/data
```
Comment (maciejgrzybek) [inline presto-cassandra/src/test/resources/cu-cassandra.yaml]: How is it related to `cassandra-unit` dependency (mentioned in the commit message)?

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSession.java (hunk 1 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSession.java]: It takes session from cache considering schema. But in the cache loader the session was being created without taking key into account.

**Comment by arhimondr** [comment]
> Please explain the f4e97db.
> What was wrong with caching partition keys?
> How does it affect performance? Have you benchmarked it?

This cache was intended to cache binary partition keys for all the partitions we ever queried. If there are to many of them - async reload every 2 minutes kills Cassandra.

I was thinking to limit the cache size first - but then i realized that there almost no pros of caching them. There are basically 3 cases that are possible.

1. User query one single partition very frequent, and this partition is very small (few rows). In this case most likely there will be a lot of such partitions, so we will overflood our cache.

2. User query one single partition frequently, and this partition is large - then the time of fetching binary partition key is negligible comparing to the time for fetching partition rows. 

Also there is a built-in cache key in Cassandra. And Cassandra itself has been built with `no extra caching needed` in mind.

**Comment by arhimondr** [comment]
> Can you explain (maybe in commit message) why you needed b798653?
> Again, how does it affect performance?

After previous commit i think it is self explanatory. We don't need that many threads just to refresh table schema and keyspaces.

**Comment by arhimondr** [comment]
> Please add explanation to the commit (b55d388 ) message what executeQuery(schema, sql) does why it does not fire a query on that schema.

Because of the way that cache is implemented. Probably it was just a bug, or something. But anyway, we dont want to keep multiple session per Presto node. In case of large Presto installation it will lead to opening a way more connections than needed.

**Comment by arhimondr** [comment]
> What we needed the per-schema session cache for? Is it because of what was described in e49f206?

To keep one session per keyspace.

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/MockCassandraSession.java (hunk 2 above)
Comment (arhimondr) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/MockCassandraSession.java]: It is not automatic. I had to do this almost manually because of that `MockCassandraSession`

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java (hunk 4 above)
Comment (arhimondr) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java]: It must be closed only upon failure. The logic is correct here.

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/CassandraTestingUtils.java (hunk 5 above)
Comment (arhimondr) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/CassandraTestingUtils.java]: It is propagated. If it fails when creating a keyspace it will throw an exception.

File: presto-cassandra/src/test/resources/cu-cassandra.yaml (hunk 7 above)
Comment (arhimondr) [inline presto-cassandra/src/test/resources/cu-cassandra.yaml]: `cassandra-unit` was doing all the `target/embeddedCassandra` magic. I'm not so smart, so i decided to go with more obvious way, and create temprorary directory manually. If it is created manually - it must be specified manually.

**Comment by maciejgrzybek** [comment]
AD https://github.com/prestodb/presto/pull/7600#issuecomment-287209579: thanks captain obvious. ;) I was wondering why do we need one schema per keyspace.

**Comment by cawallin** [comment]
Merged, thanks!
